[Quant][fx] Define native backend_config_dict for linear and conv

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74636

Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D35225680](https://our.internmc.facebook.com/intern/diff/D35225680)

andrewor14 committed:
[Quant][fx] Define native backend_config_dict for linear and conv

Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -1,44 +1,216 @@
+    return conv_configs
+
+def get_native_backend_config_dict():
+    """ Get backend for PyTorch Native backend_config_dict (fbgemm/qnnpack). """
```
nit: swap `backend` and `backend_config_dict`?

jerryzh168 reviewed (approved):
looks great, thanks!

jerryzh168 commented:
there is some error in numeric suite, which I already fixed in my PR, should I land mine first? it might have some merge conflict with your PR

andrewor14 commented:
Yes that's fine. I was expecting conflicts when I moved the dtype_config

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/match_utils.py:
```diff
@@ -124,16 +124,21 @@ def record_match(pattern, node, matched):
 
+    # Ensure tuple patterns are matched first, e.g. match (F.relu, F.linear) before F.relu,
+    # such that we do not need to impose an ordering requirement on the patterns map
+    sorted_patterns = sorted(patterns.keys(), key=lambda k: -1 if isinstance(k, tuple) else 1)
```
This was needed to fix functional linear/conv + relu

andrewor14 replied:
This was needed to fix functional linear/conv + relu

jerryzh168 commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -92,16 +96,205 @@ def _get_default_op_backend_config(op, dtype_configs):
+    dtype_configs = [weighted_op_int8_dtype_config]
+    linear_configs = []
+
+    # linear module
```
can we classify (by adding comments) these configs a bit to make sure we have more structured support,
e.g.
1. single linear
2. linear + relu
3. linear + bn

jerryzh168 commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -92,16 +96,205 @@ def _get_default_op_backend_config(op, dtype_configs):
+    })
+    return linear_configs
+
+def _get_conv_configs():
```
same for conv I think, we can classify these configs a bit, to make sure it is easier to add things in the future.

jerryzh168 reviewed (approved):
Looks great! had some nit comments inline

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D35225680](https://our.internmc.facebook.com/intern/diff/D35225680)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225680).

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D35225680](https://our.internmc.facebook.com/intern/diff/D35225680)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225680).

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D35225680](https://our.internmc.facebook.com/intern/diff/D35225680)

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D35225680](https://our.internmc.facebook.com/intern/diff/D35225680)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225680).

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D35225680](https://our.internmc.facebook.com/intern/diff/D35225680)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225680).

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D35225680](https://our.internmc.facebook.com/intern/diff/D35225680)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225680).

jerryzh168 commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -92,16 +96,222 @@ def _get_default_op_backend_config(op, dtype_configs):
             *_DEFAULT_OP_INT8_CONFIGS,
             *_DEFAULT_OP_FP16_CONFIGS,
             *_DEFAULT_OP_INT8_OR_FP16_CONFIGS,
+            *_get_linear_configs(),
```
I think we also support fp16 linear quantization, how did we support them here in the config?

jerryzh168 replied:
https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/_lower_to_native_backend.py#L286

andrewor14 replied:
(As discussed offline, these will be covered in a future PR)

andrewor14 committed:
Update on "[Quant][fx] Define native backend_config_dict for linear and conv"


Summary: This commit changes how quantization patterns for linear
and conv are set up in prepare. Previously, these were set up
through ConvReluQuantizeHandler and LinearReLUQuantizeHandler.
After this commit, however, these were set up through the
corresponding entries in the native backend_config_dict,
rendering the above quantize handlers no longer necessary.
In future commits, we will do the same for the remaining ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryzh168

Differential Revision: [D35225680](https://our.internmc.facebook.com/intern/diff/D35225680)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35225680).
